include/oneapi/dpl/internal/random_impl/counter_based_engines_stuff.h

```diff
@@ -0,0 +1,99 @@
+// -*- C++ -*-
+//===-- counter_based_engines_stuff.h -------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Abstract:
+//
+// Header file provides stuff functionality for counter-based RNG engines.
+
+#ifndef _ONEDPL_CTR_ENGINES_STUFF_H
+#define _ONEDPL_CTR_ENGINES_STUFF_H
+
+#include <concepts>
+#include <iterator>
+#include <utility>
+#include <tuple>
+#include <limits>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace internal
+{
+namespace experimental
+{
+
+/* word_mask<_U, __W> - an unsigned integral type with the low __W bits set */
+template <typename _U, unsigned __W,
+          typename = ::std::enable_if_t<::std::is_unsigned_v<_U> && (__W <= ::std::numeric_limits<_U>::digits)>>
+constexpr _U word_mask = __W ? (_U(~(_U(0))) >> (::std::numeric_limits<_U>::digits - __W)) : 0;
+
+/* For unpacking variadic of constants into two arrays */
+template <typename _UIntType, typename _Tuple, ::std::size_t... __Is>
+constexpr auto
+get_even_array_from_tuple(_Tuple __t, ::std::index_sequence<__Is...>)
+{
+    return ::std::array<_UIntType, ::std::index_sequence<__Is...>::size()>{::std::get<__Is * 2>(__t)...};
+}
+template <typename _UIntType, typename _Tuple, ::std::size_t... __Is>
+constexpr auto
+get_odd_array_from_tuple(_Tuple __t, std::index_sequence<__Is...>)
+{
+    return ::std::array<_UIntType, ::std::index_sequence<__Is...>::size()>{::std::get<__Is * 2 + 1>(__t)...};
+}
+
+/* Implement __W-bit mulhilo - returns the __W hi and __W low 
+   bits of the 2__W-bit product of __a and __b */
+template <typename _UIntType, unsigned __W>
```

Comment:
> please, align amount of "_" in templates


Initially I followed the logic `__` for values and `_` for types, so the amount of underscores may vary inside one template. But no problem, I've aligned amount of "_" in all templates.(_NOTE:_ I had to add two underscores for I/O methods, otherwise there was a problem of template parameters shadowing)

> _W should be std::size_t

I moved `mulhilo` into Philox class, so this template param is not necessary anymore.